torch/_functorch/aot_autograd.py

```diff
@@ -861,181 +876,145 @@ def maybe_to_fresh_input(idx, t, meta):
         return t.view(t.shape)
     return t
 
-# This function takes in a forward fn, runs it, and (optionally) runs autograd to compute the joint.
-# When maybe_tangents is None, we only run the forward. Otherwise we run the "joint" forward + backward.
-# Preconditions:
-# - fn corresponds to the flattened user fw function, with duplicate inputs removed
-# - functionalization is turned on (and inputs are wrapped in functional tensors)
-# - Synthetic bases have been *removed* (we've taken views on them corresponding to the user argument views).
-# - primals_after_cloning are what we run our forward function on. It is identical to primals_before_cloning,
-#   except that every input we know will be mutated in the forward has been cloned.
-#   We run our forward on primals_after_cloning (potentially mutating some inputs), and then compute our gradients
-#   w.r.t. primals_before_cloning (so we properly capture the mutation in our gradient computation).
+def fn_no_extra_mutations(
+    fn: Callable,
+    meta: ViewAndMutationMeta,
+    keep_input_mutations: bool,
+) -> Any:
+    def inner_fn(*args):
+        outs = fn(*args)
+        assert len(meta.output_info) == len(outs)
+        # The compiled fw will return mutated input tensors, *including* metadata-only mutation.
+        # However, if keep_input_mutations is set, the compiled fw only needs to return metadata-mutated inputs.
+        # (because data-only input mutations are handled directly in the compiled graph)
+        mutated_inputs_to_return = [
+            x
+            for (i, x) in enumerate(args)
+            if meta.input_info[i].mutates_metadata or (meta.input_info[i].mutates_data and not keep_input_mutations)
+        ]
+        return *mutated_inputs_to_return, *outs
+    return inner_fn
+
+# This function takes in a forward fn, and constructs a joint fwd + bwd.
 # Importantly, due functionalization + some autograd.Function constraints, this function can return EXTRA outputs
 # compared to what the original user forward returns.
-#
-# If we are only running the forward (and not computing the joint):
-# - Our function will return (updated_inputs, fw_outs)
-#
-# If we are running the forward + backward (computing the joint):
-# - Our function will return (updated_inputs, fw_outs, intermediate_bases), (gradients)
-#
-# Finally, if keep_input_mutations is set, then we will explicitly *not* return updated inputs, for any inputs
+# Our function will return (updated_inputs, fw_outs, intermediate_bases), (gradients)
+# If keep_input_mutations is set, then we will explicitly *not* return updated inputs, for any inputs
 # that experienced data-only mutations.
-# Instead, we are relying on the logic in create_forward_or_joint_functionalized to manually perform the input mutations,
+# Instead, we are relying on the logic in functionalize_fn to manually perform the input mutations,
 # keeping them directly in the traced graph.
-def forward_or_joint(
+def joint_fn_no_external_mutations(
```

Comment:
What is meant by "no external mutations" here?

Reply:
I'm going to rewrite this to have:

(1) `joint_fn()` be its own thing (that just runs function, and performs autograd.grad on the outputs that require gradients)

(2) a separate function that handles all of the input mutation / cloning / intermediate base handling. That's sort of what I meant by the "no_external_mutations" bit, although it's not a great name. For now locally, I named my function `new_fn = fn_prepped_for_autograd(fn, fw_metadata)`